core/src/solution_submission.rs

```diff
@@ -217,28 +225,59 @@ async fn retry_with_gas_price_increase(
                 solution.clone(),
                 claimed_objective_value,
                 gas_price,
-                if gas_price == gas_cap {
-                    None
-                } else {
-                    Some(BLOCK_TIMEOUT)
-                },
+                Some(BLOCK_TIMEOUT),
             )
             .await;
 
+        let before_deadline = deadline
+            .map(|deadline| Instant::now() < deadline)
+            .unwrap_or(true);
         let gas_price_can_still_increase = gas_price < effective_gas_cap;
+
         // Breaking condition for our loop
-        gas_price_can_still_increase
-            && matches!(
-                result,
-                Err(MethodError {
-                    inner: ExecutionError::ConfirmTimeout,
-                    ..
-                })
-            )
+        match result {
+            Err(MethodError {
+                inner: ExecutionError::ConfirmTimeout,
+                ..
+            }) => {
+                if gas_price_can_still_increase && before_deadline {
+                    true
+                } else {
+                    let gas_price = if gas_price_can_still_increase {
+                        gas_price_cap
+                    } else {
+                        U256::from(
+                            (gas_price.as_u128() as f64 * min_gas_price_increase_factor).ceil()
+                                as u128,
+                        )
+                    };
+                    log::info!(
```

Comment:
Are we cancelling the tx if the gas price can no longer increase? I think in that case we should hope that the current gas price is high enough to complete before the deadline and only cancel after the deadling.